Sprint 7 solution

Naz1anmak opened the pull request:
- Добавлена реализация автосохранения задач
- Изменены модификаторы доступа в InMemoryTaskManager
- Добавлены новые тесты

Naz1anmak committed:
Version 4.1
- Добавлена реализация автосохранения задач
- Изменены модификаторы доступа в InMemoryTaskManager
- Добавлены новые тесты

afomkina commented on src/main/Main.java:
```diff
@@ -92,4 +107,52 @@ public static void main(String[] args) {
+
+    public static void case2() {
+        System.out.println(" ".repeat(18) + "***  Проверка FileBackedTaskManager  ***");
+        /*Path saveFilePath = Paths.get("src/history/autoSave.csv");
```
Закомментированный код нужно удалить

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,327 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class FileBackedTaskManager extends InMemoryTaskManager implements TaskManager {
```
`InMemoryTaskManager` уже реализует `TaskManager`, поэтому явно его реализовывать в данном классе не надо

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,327 @@
+        try {
+            if (autoSavePath == null || !Files.exists(autoSavePath) || !Files.isRegularFile(autoSavePath)) {
+                Path dir = Paths.get("src/history");
+                Path defaultFile = dir.resolve("autoSave.csv");
```
Строки `"src/history"` и `"autoSave.csv"` нужно сделать приватными константами

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,327 @@
+            } else {
+                this.autoSaveFile = autoSavePath.toFile();
+            }
+        } catch (IOException e) {
```
Нужно переименовать все однобуквенные переменные

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,327 @@
+            throw new RuntimeException("Ошибка при создании файла для автосохранения.", e);
+        }
+
+        if (autoSaveFile.exists()) {
```
Файл может существовать, но быть пустым, только пользователь должен решать, когда он хочет загрузить данные из файла, а когда нет. Нужно удалить данный блок кода

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,327 @@
+        FileBackedTaskManager fileBackedTaskManager = new FileBackedTaskManager(file);
+
+        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
+            while (br.ready()) {
```
Нужно переименовать все сокращения и однобуквенные переменные

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,327 @@
+                if (!s.isEmpty()) {
+                    Task task = fileBackedTaskManager.fromString(s);
+
+                    if (task instanceof Epic) {
```
Нужно избавиться от `instanceof`

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,327 @@
+                String s = br.readLine().trim();
+
+                if (!s.isEmpty()) {
+                    Task task = fileBackedTaskManager.fromString(s);
```
 `Task` может быть `null`, строит проверить, что он не `null` прежде чем выполнять с ним какие-либо действия

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,327 @@
+                epic.getId(), TasksTypes.EPIC, epic.getName(), epic.getStatus(), epic.getDescription());
+    }
+
+    private String toString(Subtask subtask) {
```
Методы со строк 102 - 112 лучше перенести в соответсвующие объекты `Task`, `Epic` и `Sabtask`, и переопределить встроенные методы `toString`. Так мы сократим код данного менеджера и оставим только части за которые он ответсвенен

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,327 @@
+        String[] fields = value.split(",");
+        TasksTypes tasksTypes = TasksTypes.valueOf(fields[1]);
+
+        switch (tasksTypes) {
```
Лучше переписать с помощью [switch Expressions](https://www.baeldung.com/java-switch#:~:text=5.-,switch%20Expressions,-JDK%2013%20is)

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,327 @@
+    }
+
+    @Override
+    public List<Subtask> getSubtasks() {
```
Метод идентичен тому, что есть в классе родителе. Его переопределять не нужно

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,327 @@
+    }
+
+    @Override
+    public List<Epic> getEpics() {
```
Метод идентичен тому, что есть в классе родителе. Его переопределять не нужно

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,327 @@
+    }
+
+    @Override
+    public List<Task> getTasks() {
```
Метод идентичен тому, что есть в классе родителе. Его переопределять не нужно

afomkina commented on test/FileBackedTaskManagerTest.java:
папку `test` нужно перенести внутрь папки `src`

Naz1anmak replied:
Build test на гитхабе не пропускают из-за того, что test не отдельная директория.

afomkina commented on test/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,90 @@
+
+public class FileBackedTaskManagerTest {
+    File file = null;
+    FileBackedTaskManager fileBackedTaskManager = null;
```
У всех переменных класса должен быть модификатор дотупа

afomkina commented on test/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @Test
+    void savingAndLoadingTasks() {
```
Этот тестовый метод слишном большой. Нужно разбить его на несколько подтестов. При написании тестов в коммерческой разработке используют правило 1 тест - 1 тест кейс, чтобы было их проще поддерживать и выявлять ошибки

Naz1anmak committed:
Version 4.1.1
- Тесты нового класса разделены на несколько методов, переработана реализация
- Изменены методы toString у всех классов задач
- Исправлены модификаторы доступа и другие ошибки

Naz1anmak committed:
Version 4.1.2

Naz1anmak committed:
Version 4.1.2

Naz1anmak committed:
Version 4.1.3

Naz1anmak committed:
Version 4.1.3

Naz1anmak committed:
Version 4.1.4

Naz1anmak committed:
Version 4.1.5

Naz1anmak committed:
Version 4.1.6
- Обновление README.md
